File: 1209-remove-all-adjacent-duplicates-in-string-ii/1209-remove-all-adjacent-duplicates-in-string-ii.cpp
class Solution {
public:
    string removeDuplicates(string s, int k) {
        
        stack<pair<char,int>> st;
        for(char x : s) {
            if(st.empty() or st.top().first != x) {
                st.push({x,1});
            }
            else {
                st.top().second += 1;
                if(st.top().second == k) {
                    st.pop();
                }
            }
        }
        s = "";
        while(!st.empty()) {
            auto t = st.top();
            st.pop();
            for(int i = 0; i < t.second; i++) {
                s += t.first;
            }
        }
        reverse(s.begin(),s.end());
        return s;
    }
};

File: 1209-remove-all-adjacent-duplicates-in-string-ii/1209-remove-all-adjacent-duplicates-in-string-ii.py
class Solution:
    def removeDuplicates(self, s: str, k: int) -> str:
        stack = []
        for i in range(len(s)):
            if len(stack) == 0:
                stack.append([s[i],1])
            else:
                if(stack[-1][0] == s[i]):
                    stack.append([s[i],stack[-1][1] + 1])
                else:
                    stack.append([s[i],1])
                    
                if(stack[-1][1] == k):
                    stack = stack[:len(stack) - k]
                    countCurr = 0
        
        ans = ""
        for x in stack:
            ans += x[0]
        
        return ans
        